importlib_metadata/__init__.py

```diff
@@ -647,10 +699,7 @@ def entry_points():
     :return: EntryPoint objects for all installed packages.
     """
     eps = itertools.chain.from_iterable(dist.entry_points for dist in distributions())
-    by_group = operator.attrgetter('group')
-    ordered = sorted(eps, key=by_group)
-    grouped = itertools.groupby(ordered, by_group)
-    return {group: tuple(eps) for group, eps in grouped}
```

Comment:
if you back out this portion of the patch I have no opposition -- keeping this a dictionary makes a lot of common operations easier

Reply:
I'm trying to serve two use-cases here, the very common access by group, but the also not uncommon access to all entrypoints irrespective of group. I see the list of all entry points as the lowel-level interface and the mapping as a more useful but higher-level interface. By providing only the higher-level interface, it becomes clumsy and expensive to derive the list of all entry points from the mapping of entry points by group. It also becomes impossible to restore the original order of entrypoints. One use-case I'd like to address (I forget where it was requested) is to allow a tool to simply enumerate all entry points.

If we're going to go to the trouble of tweaking the interface, I'd like to make it more general and useful and regular.

By wrapping the result in a GroupedEntryPoints, one gets the generality of the list of entry points but the convenience of accessing them by group (like the dict construction allowed). Moreover, users are given an explicit interface to support these use-cases rather inspect the code to find an approach that happens to work (e.g. `all_entry_points = list(itertools.chain.from_iterable(entry_points().values()))`).

Moreover, `GroupedEntryPoints` has the additional advantage of supplying the empty set when no packages present the group, making flake8's usage more straightforward as well:

```diff
- eps = importlib_metadata.entry_points().get(self.namespace, ())
+ eps = importlib_metadata.entry_points()[self.namespace]
```

Reply:
to be clear, I don't care if the resulting code "looks cleaner".  if I have to deal with users complaining about a DeprecationWarning because you decided an api can be more clever I'm not happy.

these cutesy types that half imitate a dict but aren't a dict and a list but are neither *are going to confuse users* and by switching to them you cause me and my users pain

Reply:
The latest code no longer tries to half imitate a dict, but simply uses `__getitem__` for entry point access by name.